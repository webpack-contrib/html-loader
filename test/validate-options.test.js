import { getCompiler, compile } from './helpers';

describe('validate options', () => {
  const tests = {
    attributes: {
      success: [
        true,
        false,
        {
          list: [
            {
              tag: 'img',
              attribute: 'src',
              type: 'src',
            },
          ],
        },
        {
          list: [
            {
              tag: 'img',
              attribute: 'src',
              type: 'src',
              filter: () => true,
            },
          ],
        },
        {
          list: [
            {
              tag: 'img',
              attribute: 'src',
              type: 'src',
            },
            {
              tag: 'img',
              attribute: 'srcset',
              type: 'srcset',
            },
          ],
        },
        { urlFilter: () => true },
        { root: '.' },
        {
          list: [
            {
              tag: 'img',
              attribute: 'src',
              type: 'src',
            },
            {
              tag: 'img',
              attribute: 'srcset',
              type: 'srcset',
            },
          ],
          urlFilter: () => true,
          root: '.',
        },
      ],
      failure: [
        'true',
        [],
        {
          list: [],
        },
        {
          list: [
            {
              tag: 'img',
              attribute: 'src',
            },
          ],
        },
        {
          list: [
            {
              tag: 'img',
              attribute: 'src',
              type: 'src',
            },
            {
              tag: 'img',
              attribute: 'src',
              type: 'src',
            },
          ],
        },
        {
          list: [
            {
              tag: 'img',
              attribute: 'src',
              type: 'src',
              filter: 'test',
            },
          ],
        },
        { urlFilter: false },
        { root: false },
        { unknown: true },
      ],
    },
    esModule: {
      success: [true, false],
      failure: ['true'],
    },
    unknown: {
      success: [],
      failure: [1, true, false, 'test', /test/, [], {}, { foo: 'bar' }],
    },
  };

  function stringifyValue(value) {
    if (
      Array.isArray(value) ||
      (value && typeof value === 'object' && value.constructor === Object)
    ) {
      return JSON.stringify(value);
    }

    return value;
  }

  async function createTestCase(key, value, type) {
    it(`should ${
      type === 'success' ? 'successfully validate' : 'throw an error on'
    } the "${key}" option with "${stringifyValue(value)}" value`, async () => {
      // For loaders
      const compiler = getCompiler('simple.js', { [key]: value });

      let stats;

      try {
        stats = await compile(compiler);
      } finally {
        if (type === 'success') {
          expect(stats.hasErrors()).toBe(false);
        } else if (type === 'failure') {
          const {
            compilation: { errors },
          } = stats;

          expect(errors).toHaveLength(1);
          expect(() => {
            throw new Error(errors[0].error.message);
          }).toThrowErrorMatchingSnapshot();
        }
      }
    });
  }

  for (const [key, values] of Object.entries(tests)) {
    for (const type of Object.keys(values)) {
      for (const value of values[type]) {
        createTestCase(key, value, type);
      }
    }
  }
});
