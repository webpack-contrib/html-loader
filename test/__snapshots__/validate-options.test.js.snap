// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate options should throw an error on the "esModule" option with "true" value 1`] = `
"Invalid options object. HTML Loader has been initialized using an options object that does not match the API schema.
 - options.esModule should be a boolean.
   -> Enable or disable ES modules syntax.
   -> Read more at https://github.com/webpack-contrib/html-loader#esmodule"
`;

exports[`validate options should throw an error on the "minimize" option with "true" value 1`] = `
"Invalid options object. HTML Loader has been initialized using an options object that does not match the API schema.
 - options.minimize should be one of these:
   boolean | object { … }
   -> Tell html-loader to minimize HTML.
   -> Read more at https://github.com/webpack-contrib/html-loader#minimize
   Details:
    * options.minimize should be a boolean.
    * options.minimize should be an object:
      object { … }"
`;

exports[`validate options should throw an error on the "preprocessor" option with "true" value 1`] = `
"Invalid options object. HTML Loader has been initialized using an options object that does not match the API schema.
 - options.preprocessor should be an instance of function.
   -> Allows pre-processing of content before handling.
   -> Read more at https://github.com/webpack-contrib/html-loader#preprocessor"
`;

exports[`validate options should throw an error on the "sources" option with "[]" value 1`] = `
"Invalid options object. HTML Loader has been initialized using an options object that does not match the API schema.
 - options.sources should be one of these:
   boolean | object { list?, urlFilter?, scriptingEnabled? }
   -> By default every loadable attributes (for example - <img src='image.png'>) is imported (const img = require('./image.png'). You may need to specify loaders for images in your configuration.
   -> Read more at https://github.com/webpack-contrib/html-loader#sources
   Details:
    * options.sources should be a boolean.
    * options.sources should be an object:
      object { list?, urlFilter?, scriptingEnabled? }"
`;

exports[`validate options should throw an error on the "sources" option with "{"list":[]}" value 1`] = `
"Invalid options object. HTML Loader has been initialized using an options object that does not match the API schema.
 - options.sources.list should be a non-empty array."
`;

exports[`validate options should throw an error on the "sources" option with "{"list":[{"tag":"","attribute":"src","type":"src"}]}" value 1`] = `
"Invalid options object. HTML Loader has been initialized using an options object that does not match the API schema.
 - options.sources.list[0].tag should be a non-empty string."
`;

exports[`validate options should throw an error on the "sources" option with "{"list":[{"tag":"img","attribute":"src","type":"src","filter":"test"}]}" value 1`] = `
"Invalid options object. HTML Loader has been initialized using an options object that does not match the API schema.
 - options.sources.list[0].filter should be an instance of function."
`;

exports[`validate options should throw an error on the "sources" option with "{"list":[{"tag":"img","attribute":"src","type":"src"},{"tag":"img","attribute":"src","type":"src"}]}" value 1`] = `
"Invalid options object. HTML Loader has been initialized using an options object that does not match the API schema.
 - options.sources.list should not contain the item '[object Object]' twice."
`;

exports[`validate options should throw an error on the "sources" option with "{"list":[{"tag":"img","attribute":"src"}]}" value 1`] = `
"Invalid options object. HTML Loader has been initialized using an options object that does not match the API schema.
 - options.sources.list[0] misses the property 'type'. Should be:
   "src" | "srcset""
`;

exports[`validate options should throw an error on the "sources" option with "{"scriptingEnabled":"true"}" value 1`] = `
"Invalid options object. HTML Loader has been initialized using an options object that does not match the API schema.
 - options.sources.scriptingEnabled should be a boolean."
`;

exports[`validate options should throw an error on the "sources" option with "{"unknown":true}" value 1`] = `
"Invalid options object. HTML Loader has been initialized using an options object that does not match the API schema.
 - options.sources has an unknown property 'unknown'. These properties are valid:
   object { list?, urlFilter?, scriptingEnabled? }"
`;

exports[`validate options should throw an error on the "sources" option with "{"urlFilter":false}" value 1`] = `
"Invalid options object. HTML Loader has been initialized using an options object that does not match the API schema.
 - options.sources.urlFilter should be an instance of function."
`;

exports[`validate options should throw an error on the "sources" option with "true" value 1`] = `
"Invalid options object. HTML Loader has been initialized using an options object that does not match the API schema.
 - options.sources should be one of these:
   boolean | object { list?, urlFilter?, scriptingEnabled? }
   -> By default every loadable attributes (for example - <img src='image.png'>) is imported (const img = require('./image.png'). You may need to specify loaders for images in your configuration.
   -> Read more at https://github.com/webpack-contrib/html-loader#sources
   Details:
    * options.sources should be a boolean.
    * options.sources should be an object:
      object { list?, urlFilter?, scriptingEnabled? }"
`;

exports[`validate options should throw an error on the "unknown" option with "/test/" value 1`] = `
"Invalid options object. HTML Loader has been initialized using an options object that does not match the API schema.
 - options has an unknown property 'unknown'. These properties are valid:
   object { preprocessor?, sources?, minimize?, esModule? }"
`;

exports[`validate options should throw an error on the "unknown" option with "[]" value 1`] = `
"Invalid options object. HTML Loader has been initialized using an options object that does not match the API schema.
 - options has an unknown property 'unknown'. These properties are valid:
   object { preprocessor?, sources?, minimize?, esModule? }"
`;

exports[`validate options should throw an error on the "unknown" option with "{"foo":"bar"}" value 1`] = `
"Invalid options object. HTML Loader has been initialized using an options object that does not match the API schema.
 - options has an unknown property 'unknown'. These properties are valid:
   object { preprocessor?, sources?, minimize?, esModule? }"
`;

exports[`validate options should throw an error on the "unknown" option with "{}" value 1`] = `
"Invalid options object. HTML Loader has been initialized using an options object that does not match the API schema.
 - options has an unknown property 'unknown'. These properties are valid:
   object { preprocessor?, sources?, minimize?, esModule? }"
`;

exports[`validate options should throw an error on the "unknown" option with "1" value 1`] = `
"Invalid options object. HTML Loader has been initialized using an options object that does not match the API schema.
 - options has an unknown property 'unknown'. These properties are valid:
   object { preprocessor?, sources?, minimize?, esModule? }"
`;

exports[`validate options should throw an error on the "unknown" option with "false" value 1`] = `
"Invalid options object. HTML Loader has been initialized using an options object that does not match the API schema.
 - options has an unknown property 'unknown'. These properties are valid:
   object { preprocessor?, sources?, minimize?, esModule? }"
`;

exports[`validate options should throw an error on the "unknown" option with "test" value 1`] = `
"Invalid options object. HTML Loader has been initialized using an options object that does not match the API schema.
 - options has an unknown property 'unknown'. These properties are valid:
   object { preprocessor?, sources?, minimize?, esModule? }"
`;

exports[`validate options should throw an error on the "unknown" option with "true" value 1`] = `
"Invalid options object. HTML Loader has been initialized using an options object that does not match the API schema.
 - options has an unknown property 'unknown'. These properties are valid:
   object { preprocessor?, sources?, minimize?, esModule? }"
`;
